bionic/flow.py

```diff
@@ -1231,9 +1231,21 @@ def get(self, name, collection=None, fmt=None, mode=object):
             return set(values)
         elif collection is pd.Series or collection == "series":
             if len(result_group.key_space) > 0:
+                providers_by_name = {
+                    case_name: self._state.get_provider(case_name)
+                    for case_name in list(result_group.key_space)
+                }
+                for case_name, provider in providers_by_name.items():
+                    if not isinstance(provider, ValueProvider):
+                        message = f"""
+                        Provider for the entity {case_name} was expected to be
+                        a ValueProvider but wasn't.
+                        """
+                        raise AssertionError(oneline(message))
                 index = multi_index_from_case_keys(
                     case_keys=[result.query.case_key for result in result_group],
                     ordered_key_names=list(result_group.key_space),
+                    providers_by_name=providers_by_name,
```

Comment:
The way we're pulling the values directly out of the ValueProviders is a little awkward, and it will be even more awkward when we have tuple descriptors. Instead, what if we used `self._deriver.derive()` here to get the values for each of entities in the key space, and then use those values to construct the index? Then we wouldn't need as many changes to `ValueProvider`.

Reply:
This definitely sounds neater to me but I don't really understand how to implement this. How are we going to match the case key of `name` in `Flow.get(name)` with the case keys of the result group derived for the entity names in key space? The case keys for them can be different.

Reply:
Here is my thinking (but it's possible I've missed something):

Each name in the original entity's key space should be the name of an ancestor entity, and that ancestor entity's key space should be a subset of the original's. So to find the ancestor value that corresponds to the original value, we should just need to find the ancestor case key whose values match the original within that subset.

If that's still confusing, want to chat by phone?

Reply:
I'll see how hard it is to implement a is_subset / is_superset methods on `CaseKey`. Otherwise, this makes sense! Thank you!

(Rebased the branch to pickup github actions updates)

Reply:
Doh, `CaseKey` and `CaseKeySpace` already support set operations that I need (in the form of `select` and `project`). Thanks Janek for pointing out that set operations exist!

Reply:
Done!